db/background_mgr.go

```diff
@@ -0,0 +1,200 @@
+//  Copyright 2012-Present Couchbase, Inc.
+//
+//  Use of this software is governed by the Business Source License included
+//  in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+//  in that file, in accordance with the Business Source License, use of this
+//  software will be governed by the Apache License, Version 2.0, included in
+//  the file licenses/APL2.txt.
+
+package db
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/couchbase/sync_gateway/base"
+)
+
+type BackgroundProcessState string
+
+const (
+	BackgroundProcessStateRunning  BackgroundProcessState = "running"
+	BackgroundProcessStateStopped  BackgroundProcessState = "stopped"
+	BackgroundProcessStateStopping BackgroundProcessState = "stopping"
+	BackgroundProcessStateError    BackgroundProcessState = "error"
+)
+
+type BackgroundProcessAction string
+
+const (
+	BackgroundProcessActionStart BackgroundProcessAction = "start"
+	BackgroundProcessActionStop  BackgroundProcessAction = "stop"
+)
+
+// BackgroundManagerI this is an interface satisfied by BackgroundManager
+// This is technically not needed, however, it allows easily visibility into exposed methods
+type BackgroundManagerI interface {
+	Start(options map[string]interface{})
+	GetStatus() []byte
+	ResetStatus()
+	Stop()
+}
+
+// BackgroundManager this is the over-arching type which is exposed in DatabaseContext
+type BackgroundManager struct {
+	BackgroundManagerStatus
+	terminator chan struct{}
+	lock       sync.Mutex
+	Process    BackgroundManagerProcessI
+}
+
+// BackgroundManagerStatus simply stores data used in BackgroundManager. This data can also be exposed to users over
+// REST. Splitting this out into an additional embedded struct allows easy JSON marshalling
+type BackgroundManagerStatus struct {
+	State     BackgroundProcessState `json:"status"`
+	LastError error                  `json:"last_error"`
+}
+
+// BackgroundManagerProcessI is an interface satisfied by any of the background processes
+// Examples of this: ReSync, Compaction
+type BackgroundManagerProcessI interface {
+	Run(options map[string]interface{}, terminator chan struct{}) error
+	GetProcessStatus() []base.KVPair
+	ResetStatus()
+}
+
+var _ BackgroundManagerI = &BackgroundManager{}
+
+func (b *BackgroundManager) Start(options map[string]interface{}) {
+	go func() {
+		defer b.SetRunState(BackgroundProcessStateStopped)
+		err := b.Process.Run(options, b.terminator)
+		if err != nil {
+			base.Errorf("Error")
+			b.SetError(err)
+		}
+	}()
+}
+
+func (b *BackgroundManager) GetStatus() []byte {
+	b.lock.Lock()
+	status, err := base.JSONMarshal(b.BackgroundManagerStatus)
+	if err != nil {
+		status = []byte("{}")
+	}
+	b.lock.Unlock()
+
+	processPairs := b.Process.GetProcessStatus()
+
+	retStatus, err := base.InjectJSONProperties(status, processPairs...)
+	if err != nil {
+		return nil
+	}
```

Comment:
I think I'd prefer for GetProcessStatus to pass BackgroundManagerStatus down to the process, and let the process build the marshalled JSON in the usual way.  That would allow the process to omit State or LastError (if that made sense for a given use case), but more importantly would use standard json struct annotations, avoiding the need for 'test only' definitions of those.